### PR TITLE
Issue 488: use amazonka and http-client-tls

### PR DESCRIPTION
**Pull request by tiago-loureiro**
This PR introduces amazonka for cargohold too. The non-multipart tests can also be run against a local fake-s3 and cloudfront is now optional as well. OpenSSL was also removed and replaced by http-client-tls or cryptonite, depending on the usage. There are some legacy endpoints that are hard to test since there is only a way to download those assets, not to upload.

Still to be done:
- [x] Remove all references to the old AWS library (currently used by the legacy endpoints)
- [x] General cleanup

Optionally, but highly likely not in this PR:
~~- [ ] Try to get fake S3 to work for resumable uploads~~
~~- [ ] The AWS module is pretty much the same as in all other services, we should probably move it to our ropes library to replace the old AWS code~~

### CODE REVIEW

**Comment by tiago-loureiro** [comment]
I think this is good to go as-is. One caveat: multipart uploads _do not_ work with fake s3 - it does not have full support unfortunately. I did not dig in too much into it so it may or may not be a lot of work to get that working that is why I chose to still default to use S3 in tests.

In order to try out fake-s3 (which also requires disabling cloudfront by passing `--disable-cloudfront` to cargohold) just look at the commented out config in the `Procfile`.

File: services/cargohold/src/CargoHold/App.hs [hunk 1]
```diff
@@ -11,14 +11,15 @@
 module CargoHold.App
     ( -- * Environment
       Env
-    , AwsEnv (..)
+    -- , AwsEnv (..)
     , newEnv
```
[the remark this answers is not in the document]
Comment (tiago-loureiro) [reply services/cargohold/src/CargoHold/App.hs]: 👍

File: services/cargohold/deb/etc/sv/cargohold/run [hunk 2]
```diff
@@ -38,4 +38,4 @@ exec chpst -u $USER \
     --aws-cloudfront-domain=${CARGOHOLD_AWS_CLOUDFRONT_DOMAIN?'unset'} \
     --aws-cloudfront-keypair-id=${CARGOHOLD_AWS_CLOUDFRONT_KEYPAIR_ID?'unset'} \
     --aws-cloudfront-private-key=${CARGOHOLD_AWS_CLOUDFRONT_PRIVATEKEY?'unset'} \
-    --max-total-bytes=${CARGOHOLD_MAX_TOTAL_BYTES?'unset'}
```
[the remark this answers is not in the document]
Comment (tiago-loureiro) [reply services/cargohold/deb/etc/sv/cargohold/run]: 👍

File: services/cargohold/src/CargoHold/App.hs [hunk 3]
```diff
@@ -91,39 +80,25 @@ newEnv o = do
...
-    return $! AwsEnv amz s3c (awsOpts^.awsS3Bucket) sig
+    ama  <- initAws o lgr mgr
+    return $ Env ama sig met lgr mgr mempty (o^.optSettings)
 
+initAws :: Opts -> Logger -> Manager -> IO AWS.Env
+initAws o l m = AWS.mkEnv l (o^.optAws.awsS3Endpoint) (o^.optAws.awsS3Bucket) m
+
+-- TODO: If we want to have more control on the cipher suite, look into
+-- https://hackage.haskell.org/package/tls-1.4.0/docs/Network-TLS.html
+-- and make use of ClientParams
```
[the remark this answers is not in the document]
Comment (tiago-loureiro) [reply services/cargohold/src/CargoHold/App.hs]: > The openssl context had a contextLoadSystemCerts, what does the http-tls do? Do the certificates loaded include the root ones necessary for AWS?

Yes, that is what happens when we use default settings (see [here](https://github.com/vincenthz/hs-connection/blob/00a8952e9cd331d3276fd49d6a2ffa90e81f27bb/Network/Connection.hs#L119)). Uses the global context which uses certificates from the [system store](https://github.com/vincenthz/hs-connection/blob/master/Network/Connection.hs#L110)

However, I just realised that by default it uses [ciphersuite_all](https://github.com/vincenthz/hs-tls/blob/master/core/Network/TLS/Extra/Cipher.hs#L180) which include RC4 ciphers which we should probably stay away from. I made a commit to [fix that](https://github.com/wireapp/wire-server/pull/203/commits/d9d25c96f709bec64531531d1170553bdee3d2ea)

File: services/cargohold/src/CargoHold/S3.hs [hunk 4]
```diff
@@ -149,24 +150,43 @@ deleteV3 (s3Key . mkKey -> key) = do
...
+          & coContentType ?~ encodeMIMEType ct
+          & coMetadataDirective ?~ MDReplace
+          & coMetadata .~ metaHeaders tok prc
+
+signedUrl :: (ToByteString p) => p -> ExceptT Error App URI
+signedUrl path = do
+    e <- view aws
+    let b = view AWS.s3Bucket e
+    now <- liftIO getCurrentTime
+    let expiresIn = Seconds 300
```
[the remark this answers is not in the document]
Comment (tiago-loureiro) [reply services/cargohold/src/CargoHold/S3.hs]: Since these URLs are self signed, they should be valid for a short period of time since there's no good way to remove access to a specific asset.

File: deploy/docker-ephemeral/docker-compose.yaml [hunk 5]
```diff
@@ -24,16 +24,23 @@ services:
     ports:
       - "6379:6379"
 
+  fake_s3:
+      image: tiagoloureiro/fake-s3:0.2.5
```
Comment (elverkilde) [inline deploy/docker-ephemeral/docker-compose.yaml]: should we put this under a Wire org instead?

File: services/cargohold/src/CargoHold/API/V3.hs [hunk 6]
```diff
@@ -51,14 +53,14 @@ import qualified Data.Conduit.Binary     as Conduit
 import qualified Data.Text.Ascii         as Ascii
 import qualified Data.Text.Lazy          as LT
 
-upload :: V3.Principal -> Source IO ByteString -> Handler V3.Asset
+upload :: V3.Principal -> Source (ResourceT IO) ByteString -> Handler V3.Asset
```
Comment (elverkilde) [inline services/cargohold/src/CargoHold/API/V3.hs]: why is this needed?

Comment (tiago-loureiro) [reply services/cargohold/src/CargoHold/API/V3.hs]: You mean the change from `IO` to `ResourceT IO`? That's the type required by `amazonka`

Comment (elverkilde) [reply services/cargohold/src/CargoHold/API/V3.hs]: :+1:

**Comment by tiago-loureiro** [comment]
OK, I have merged the latest changes from develop in case someone wants to have a look at this. For some more context, the reason why this was not merged in was that we saw a performance degradation of ~50% in some of the tests. The changes that this PR introduces and may be root cause are:

 * Inefficient use of the `amazonka` library versus the old `aws`
 * Removal of `openssl` in favour of a pure haskell `tls` implementation

**Comment by neongreen** [comment]
Long long ago I was told that it leaks memory or something (I don't remember already). What's the status of this now?

**Comment by tiago-loureiro** [comment]
> Long long ago I was told that it leaks memory or something (I don't remember already). What's the status of this now?

[This](https://github.com/wireapp/wire-server/pull/203#issuecomment-374012262) sums up the reason why we didn't merge it in back then.

I have merged 488 with the latest develop in https://github.com/wireapp/wire-server/tree/issue-488-v2 so we can have a look at that one. Note that this new branch does not seem to work well with minio (not sure yet what the cause may be) but it works well against real S3.

**Comment by neongreen** [comment]
Closing in favor of #483.
